group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,419 @@
+        currentMemberAssignment.forEach((topicId, partitions) -> {
+            if (subscriptionIds.contains(topicId)) {
+                partitions.forEach(partition -> {
+                    TopicIdPartition topicIdPartition = new TopicIdPartition(topicId, partition);
+                    if (rackInfo.useRackStrategy && rackInfo.racksMismatch(memberId, topicIdPartition)) {
+                        currentPartitionOwners.put(topicIdPartition, memberId);
+                    } else {
+                        validCurrentAssignmentList.add(topicIdPartition);
+                    }
+                });
+            } else {
+                LOG.debug("The topic " + topicId + " is no longer present in the subscribed topics list");
+            }
+        });
+
+        return validCurrentAssignmentList;
+    }
+
+    /**
+     * Allocates the unassigned partitions to unfilled members present in the same rack.
+     * Partitions with the least number of potential members in the same rack are allotted first.
+     * Members in the same rack with the least number of partitions in the target assignment
+     * are assigned partitions first.
+     */
+    private void rackAwarePartitionAssignment() {
+        // Sort partitions in ascending order by number of potential members with matching racks.
+        // Partitions with no potential members in the same rack aren't included in this list.
+        List<TopicIdPartition> sortedPartitions = rackInfo.sortPartitionsByRackMembers(unassignedPartitions);
+
+        sortedPartitions.forEach(partition -> {
+            List<String> sortedMembersWithMatchingRack = rackInfo.getSortedMembersWithMatchingRack(partition, targetAssignment);
+
+            for (String memberId : sortedMembersWithMatchingRack) {
+                if (potentiallyUnfilledMembers.containsKey(memberId) && maybeAssignPartitionToMember(memberId, partition)) {
+                    unassignedPartitions.remove(partition);
+                    break;
+                }
+            }
+        });
+    }
+
+    /**
+     * Allocates the unassigned partitions to unfilled members in a round-robin fashion.
+     *
+     * If the rack-aware strategy is enabled, partitions are attempted to be assigned back to their current owners first.
+     * This is because pure stickiness without rack matching is not considered initially.
+     *
+     * If a partition couldn't be assigned to its current owner due to the quotas OR
+     * if the rack-aware strategy is not enabled, the partitions are allocated to the unfilled members.
+     */
+    private void unassignedPartitionsRoundRobinAssignment() {
+        Queue<String> roundRobinMembers = new LinkedList<>(potentiallyUnfilledMembers.keySet());
+
+        List<TopicIdPartition> sortedPartitionsList = unassignedPartitions.stream()
+            .sorted(Comparator.comparing(TopicIdPartition::topicId).thenComparing(TopicIdPartition::partitionId))
+            .collect(Collectors.toList());
```

Comment:
I don't recall. Why do we need to sort the partitions here?

Reply:
Sorted so that partitions from different topics are distributed equally amongst all the consumers, as a bonus it's easier to predict which partitions go to which consumer

Reply:
Understood. Should we add a small comment about it? I will very likely forget again... ;)

Reply:
yep sure will add it! I'll probably forget too lol